crates/fs/private/hamt/node.rs

```diff
@@ -0,0 +1,251 @@
+        store: &B,
+    ) -> Result<Rc<Self>> {
+        let bit_index = hashquartets.next()? as usize;
+
+        // No existing values at this point.
+        if !self.bitmask[bit_index] {
+            let mut cloned = (**self).clone();
+            cloned.insert_child(bit_index, key, value);
+            return Ok(Rc::new(cloned));
+        }
+
+        let value_index = self.get_value_index(bit_index);
+        Ok(match &self.pointers[value_index] {
+            Pointer::Values(values) => {
+                let mut cloned = (**self).clone();
+                let mut values_cloned = (*values).clone();
+
+                if let Some(i) = values.iter().position(|p| p.key == key) {
+                    values_cloned[i] = Pair { key, value };
+                } else {
+                    values_cloned.insert(value_index, Pair { key, value });
+                }
+
+                cloned.pointers[value_index] = Pointer::Values(values_cloned);
+                Rc::new(cloned)
+            }
+            Pointer::Link(link) => {
+                let child = Rc::clone(link.resolve_value(store).await?);
+                let new_child = child.modify_value(hashquartets, key, value, store).await?;
+                let mut self_cloned = (**self).clone();
+                self_cloned.pointers[value_index] = Pointer::Link(Link::from(new_child));
+                Rc::new(self_cloned)
+            }
+        })
+    }
+
+    async fn search<'a, B: BlockStore>(
+        self: &'a Rc<Self>,
+        key: &K,
+        store: &B,
+    ) -> Result<Option<&'a Pair<K, V>>> {
+        let hash = &Sha3_256::generate_hash(&key);
+        let mut hashquartets = HashQuartets::new(hash);
+        self.get_value(&mut hashquartets, key, store).await
+    }
+
+    #[async_recursion(?Send)]
+    async fn get_value<'a, 'b, B: BlockStore>(
+        self: &'a Rc<Self>,
+        hashquartets: &'b mut HashQuartets,
+        key: &K,
+        store: &B,
+    ) -> Result<Option<&'a Pair<K, V>>> {
+        let bit_index = hashquartets.next()? as usize;
+
+        if !self.bitmask[bit_index] {
+            return Ok(None);
+        }
+
+        let value_index = self.get_value_index(bit_index);
+        match &self.pointers[value_index] {
+            Pointer::Values(values) => Ok(values.iter().find(|kv| key.eq(&kv.key))),
+            Pointer::Link(link) => {
+                let child = link.resolve_value(store).await?;
+                child.get_value(hashquartets, key, store).await
+            }
+        }
+    }
+}
+
+impl<K, V> Default for Node<K, V> {
+    fn default() -> Self {
+        Node {
+            bitmask: BitArray::ZERO,
+            pointers: Vec::new(),
+        }
+    }
+}
+
+impl<K, V> Serialize for Node<K, V>
+where
+    K: Serialize,
+    V: Serialize,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        (&self.bitmask.as_raw_slice(), &self.pointers).serialize(serializer)
```

Comment:
Wait, how is that even possible? I don't see a `impl Serialize for Pointers` in `pointers.rs`? :thinking: